File: iCyption2022/CftTest/harta/solution.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> tree;
int arr[100001] = {0};

void dfs(int node, int prev, int time) {
    arr[node] = time;
    for (int adj : tree[node]) {
        if (adj != prev) {
            dfs(adj, node, time+1);
        }
    }
}

int main() {
    int n; 
    cin >> n;

    int a, b, x; 
    cin >> a >> b >> x;

    tree.resize(n+1);

    for (int i = 1; i <= n-1; i++) {
        int A, B; cin >> A >> B;
        tree[A].push_back(B);
        tree[B].push_back(A);
    }

    dfs(x, -1, 0);
    int ruteA = arr[a];
    int ruteB = arr[b];

    dfs(a, -1, 0);
    ruteA += arr[b];
    ruteB += arr[b];

    cout << min(ruteA, ruteB) << '\n';
}

File: iCyption2022/CftTest/harta/solution.py
n = int(input())
a, b, x = map(int, input().split())

arr = [0 for i in range(100001)]
tree = [[] for i in range(n+1)]

def dfs(node, prev, time) :
    arr[node] = time
    for adj in tree[node]:
        if (adj != prev) :
            dfs(adj, node, time+1)

for i in range(1, n) :
    A, B = map(int, input().split())
    tree[A].append(B)
    tree[B].append(A)

dfs(x, -1, 0)
ruteA = arr[a]
ruteB = arr[b]

dfs(a, -1, 0)
ruteA += arr[b]
ruteB += arr[b]

print(min(ruteA, ruteB))